waftools/detections/compiler_swift.py

```diff
@@ -19,50 +21,162 @@ def __add_swift_flags(ctx):
+        ctx.env.SWIFT_FLAGS.extend([ "-swift-version", "5" ])
 
     if ctx.is_debug_build():
         ctx.env.SWIFT_FLAGS.append("-g")
 
     if ctx.is_optimization():
         ctx.env.SWIFT_FLAGS.append("-O")
 
-def __add_swift_library_linking_flags(ctx, swift_library):
+def __add_static_swift_library_linking_flags(ctx, swift_library):
     ctx.env.append_value('LINKFLAGS', [
         '-L%s' % swift_library,
         '-Xlinker', '-force_load_swift_libs', '-lc++',
     ])
 
+def __add_dynamic_swift_library_linking_flags(ctx, swift_library):
+    ctx.env.append_value('LINKFLAGS', [ '-L%s' % swift_library ])
+
+    #ABI compatibility
+    if StrictVersion(ctx.env.SWIFT_VERSION) >= StrictVersion("5.0"):
+        ctx.env.append_value('LINKFLAGS', [
+            '-Xlinker', '-rpath', '-Xlinker', '/usr/lib/swift',
+        ])
+
+    ctx.env.append_value('LINKFLAGS', [
+        '-Xlinker', '-rpath', '-Xlinker', swift_library,
+    ])
+
 def __find_swift_library(ctx):
-    swift_library_paths = [
-        'Toolchains/XcodeDefault.xctoolchain/usr/lib/swift_static/macosx',
-        'usr/lib/swift_static/macosx'
-    ]
+    swift_libraries = {}
+    #look for set lib paths in passed environment variables
+    if 'SWIFT_LIB_DYNAMIC' in ctx.environ:
+        swift_libraries['SWIFT_LIB_DYNAMIC'] = ctx.environ['SWIFT_LIB_DYNAMIC']
+    if 'SWIFT_LIB_STATIC' in ctx.environ:
+        swift_libraries['SWIFT_LIB_STATIC'] = ctx.environ['SWIFT_LIB_STATIC']
+
+    #search for swift libs relative to the swift compiler executable
+    swift_library_relative_paths = {
+        'SWIFT_LIB_DYNAMIC': '../../lib/swift/macosx',
+        'SWIFT_LIB_STATIC': '../../lib/swift_static/macosx'
+    }
+
+    for lib_type, path in swift_library_relative_paths.items():
+        if lib_type not in swift_libraries:
+            lib_path = os.path.join(ctx.env.SWIFT, path)
+            swift_library = ctx.root.find_dir(lib_path)
+            if swift_library is not None:
+                swift_libraries[lib_type] = swift_library.abspath()
+
+    #fall back to xcode-select path
+    swift_library_paths = {
+        'SWIFT_LIB_DYNAMIC': [
+            'Toolchains/XcodeDefault.xctoolchain/usr/lib/swift/macosx',
+            'usr/lib/swift/macosx'
+        ],
+        'SWIFT_LIB_STATIC': [
+            'Toolchains/XcodeDefault.xctoolchain/usr/lib/swift_static/macosx',
+            'usr/lib/swift_static/macosx'
+        ]
+    }
     dev_path = __run(['xcode-select', '-p'])[1:]
 
-    ctx.start_msg('Checking for Swift Library')
-    for path in swift_library_paths:
-        swift_library = ctx.root.find_dir([dev_path, path])
-        if swift_library is not None:
-            ctx.end_msg(swift_library.abspath())
-            __add_swift_library_linking_flags(ctx, swift_library.abspath())
-            return
-    ctx.end_msg(False)
+    for lib_type, paths in swift_library_paths.items():
+        for path in paths:
+            if lib_type not in swift_libraries:
```

Comment:
Since this does not depend on the paths themselves, it could just be put on the top of the loop as an early continue:
```python
for lib_type, paths in swift_library_paths.items():
    if lib_type in swift_libraries:
        continue

    for path in paths:
        ...
```